crates/core/src/nats/client.rs

```diff
@@ -0,0 +1,262 @@
+use std::{sync::Arc, time::Duration};
+
+use async_nats::connection;
+use bytes::Bytes;
+use tracing::info;
+
+use super::{
+    types::{
+        JetStreamConfig,
+        JetStreamContext,
+        NatsConsumer,
+        NatsStream,
+        PullConsumerConfig,
+    },
+    NatsError,
+    Subject,
+};
+use crate::types::BoxedResult;
+
+#[derive(Debug, Clone)]
+pub struct NatsClient {
+    pub url: String,
+    pub conn_id: String,
+    pub conn: Arc<async_nats::Client>,
+    pub jetstream: Arc<JetStreamContext>,
+}
+
+impl NatsClient {
+    pub async fn connect(
+        url: &str,
+        conn_id: &str,
+        nkey: Option<String>,
+    ) -> Result<Self, NatsError> {
+        let conn_id = conn_id.to_string();
+        let conn = Arc::new(Self::create_conn(url, nkey).await?);
+        let context =
+            Arc::new(async_nats::jetstream::new(conn.as_ref().clone()));
```

Comment:
Why do we want the `NatsClient` to be clonable?

Reply:
There's a bug related to the `jetstream::new(client)` method that just worked when using like this, I tried many different approaches and in the end this was the one that worked 😕